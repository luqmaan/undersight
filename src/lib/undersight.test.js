import {
  getTopScores,
  getTopFour,
  getRolePicks,
  getHeroCounters,
  getTeamPicksByHardCounter,
  getTeamPicksByHardCounterPrime,
  getTeamPicksByHardCounterFlexRoles,
} from './undersight';
import allCountersJSON from '../data/counters.json';
import herosJSON from '../data/heros.json';
import herosRanksJSON from '../data/heros_ranks.json';

const testCounters = allCountersJSON.filter((counter) => (
  counter.you === 'Reaper' ||
  counter.you === 'Pharah' ||
  counter.you === 'Soldier: 76'
)).filter((counter) => (
  counter.enemy === 'Reaper' ||
  counter.enemy === 'Pharah' ||
  counter.enemy === 'Soldier: 76'
));

describe('undersight', () => {
  describe('getTopScores', () => {
    it('should calculate scores', () => {
      expect(getTopScores(testCounters, ['Pharah', 'Pharah', 'Pharah'])).toMatchSnapshot();
      expect(getTopScores(testCounters, ['Reaper', 'Reaper', 'Reaper'])).toMatchSnapshot();
      expect(getTopScores(testCounters, ['Soldier: 76', 'Soldier: 76', 'Soldier: 76'])).toMatchSnapshot();
      expect(getTopScores(testCounters, ['Reaper', 'Soldier: 76', 'Pharah'])).toMatchSnapshot();
      expect(getTopScores(testCounters, ['Soldier: 76', 'Pharah', 'Reaper'])).toMatchSnapshot();
    });

    it('should calculate score for one enemy', () => {
      expect(getTopScores(testCounters, ['Pharah'])).toMatchSnapshot();
    });

    it('should calculate score for four enemy', () => {
      expect(getTopScores(testCounters, ['Pharah', 'Pharah', 'Pharah', 'Pharah'])).toMatchSnapshot();
    });
  });

  describe('getHeroCounters', () => {
    it('should return counters sorted by score desc', () => {
      expect(getHeroCounters(allCountersJSON, 'Pharah')).toMatchSnapshot();
    });
  });

  describe('getTeamPicksByHardCounter', () => {
    it('should return team picks', () => {
      expect(getTeamPicksByHardCounter(allCountersJSON, herosJSON, ['Genji', 'Junkrat', 'Roadhog', 'Zenyatta', 'Hanzo', 'Reaper'], false)).toMatchSnapshot();
    });
  });

  describe('getTeamPicksByHardCounterPrime', () => {
    it('should return team picks', () => {
      expect(getTeamPicksByHardCounterPrime(allCountersJSON, herosJSON, herosRanksJSON, ['Genji', 'Junkrat', 'Roadhog', 'Zenyatta', 'Hanzo', 'Reaper'], false)).toMatchSnapshot();
    });
  });

  describe('getTeamPicksByHardCounterFlexRoles', () => {
    it('should return team picks', () => {
      const teamPicks = ['Tracer', 'Junkrat', 'Zarya', 'Reinhardt'];
      const roleCountGoal = {tank: 3, dps: 3, support: 0};
      expect(getTeamPicksByHardCounterFlexRoles(allCountersJSON, herosJSON, herosRanksJSON, teamPicks, roleCountGoal)).toMatchSnapshot();
    });
  });
});
