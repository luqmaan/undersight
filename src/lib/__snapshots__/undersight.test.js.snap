exports[`undersight getAllRolePicks should get picks for each category 1`] = `
Object {
  "Defense": Array [
    Object {
      "name": "Hanzo",
      "score": 6
    },
    Object {
      "name": "Widowmaker",
      "score": 3
    },
    Object {
      "name": "Mei",
      "score": -6
    },
    Object {
      "name": "Torbjörn",
      "score": -6
    },
    Object {
      "name": "Junkrat",
      "score": -6
    },
    Object {
      "name": "Bastion",
      "score": -6
    }
  ],
  "Offense": Array [
    Object {
      "name": "McCree",
      "score": 6
    },
    Object {
      "name": "Soldier: 76",
      "score": 6
    },
    Object {
      "name": "Tracer",
      "score": 3
    },
    Object {
      "name": "Reaper",
      "score": 3
    },
    Object {
      "name": "Pharah",
      "score": 0
    },
    Object {
      "name": "Genji",
      "score": -3
    }
  ],
  "Support": Array [
    Object {
      "name": "Ana",
      "score": 3
    },
    Object {
      "name": "Zenyatta",
      "score": -3
    },
    Object {
      "name": "Mercy",
      "score": -3
    },
    Object {
      "name": "Symmetra",
      "score": -6
    },
    Object {
      "name": "Lúcio",
      "score": -6
    }
  ],
  "Tank": Array [
    Object {
      "name": "Roadhog",
      "score": 3
    },
    Object {
      "name": "D.Va",
      "score": 3
    },
    Object {
      "name": "Winston",
      "score": -3
    },
    Object {
      "name": "Zarya",
      "score": -6
    },
    Object {
      "name": "Reinhardt",
      "score": -6
    }
  ]
}
`;

exports[`undersight getAllRolePicks should get picks for each category 2`] = `
Object {
  "Defense": Array [
    Object {
      "name": "Widowmaker",
      "score": 9
    },
    Object {
      "name": "Hanzo",
      "score": 9
    },
    Object {
      "name": "Torbjörn",
      "score": 0
    },
    Object {
      "name": "Mei",
      "score": -3
    },
    Object {
      "name": "Junkrat",
      "score": -3
    },
    Object {
      "name": "Bastion",
      "score": -3
    }
  ],
  "Offense": Array [
    Object {
      "name": "McCree",
      "score": 12
    },
    Object {
      "name": "Tracer",
      "score": 9
    },
    Object {
      "name": "Reaper",
      "score": 9
    },
    Object {
      "name": "Soldier: 76",
      "score": 9
    },
    Object {
      "name": "Genji",
      "score": 3
    },
    Object {
      "name": "Pharah",
      "score": 3
    }
  ],
  "Support": Array [
    Object {
      "name": "Ana",
      "score": 9
    },
    Object {
      "name": "Zenyatta",
      "score": -3
    },
    Object {
      "name": "Symmetra",
      "score": -6
    },
    Object {
      "name": "Lúcio",
      "score": -6
    },
    Object {
      "name": "Mercy",
      "score": -6
    }
  ],
  "Tank": Array [
    Object {
      "name": "Roadhog",
      "score": 6
    },
    Object {
      "name": "Winston",
      "score": 0
    },
    Object {
      "name": "D.Va",
      "score": 0
    },
    Object {
      "name": "Zarya",
      "score": -9
    },
    Object {
      "name": "Reinhardt",
      "score": -9
    }
  ]
}
`;

exports[`undersight getHeroCounters should return counters sorted by score desc 1`] = `
Array [
  Object {
    "enemy": "Zarya",
    "score": 2,
    "you": "Pharah"
  },
  Object {
    "enemy": "Torbjörn",
    "score": 2,
    "you": "Pharah"
  },
  Object {
    "enemy": "Symmetra",
    "score": 2,
    "you": "Pharah"
  },
  Object {
    "enemy": "Reinhardt",
    "score": 2,
    "you": "Pharah"
  },
  Object {
    "enemy": "Mei",
    "score": 2,
    "you": "Pharah"
  },
  Object {
    "enemy": "Lúcio",
    "score": 2,
    "you": "Pharah"
  },
  Object {
    "enemy": "Junkrat",
    "score": 2,
    "you": "Pharah"
  },
  Object {
    "enemy": "Bastion",
    "score": 2,
    "you": "Pharah"
  },
  Object {
    "enemy": "Zenyatta",
    "score": 1,
    "you": "Pharah"
  },
  Object {
    "enemy": "Winston",
    "score": 1,
    "you": "Pharah"
  },
  Object {
    "enemy": "Reaper",
    "score": 1,
    "you": "Pharah"
  },
  Object {
    "enemy": "Mercy",
    "score": 1,
    "you": "Pharah"
  },
  Object {
    "enemy": "Hanzo",
    "score": 1,
    "you": "Pharah"
  },
  Object {
    "enemy": "Genji",
    "score": 1,
    "you": "Pharah"
  },
  Object {
    "enemy": "D.Va",
    "score": 1,
    "you": "Pharah"
  },
  Object {
    "enemy": "Ana",
    "score": 1,
    "you": "Pharah"
  },
  Object {
    "enemy": "Widowmaker",
    "score": 0,
    "you": "Pharah"
  },
  Object {
    "enemy": "Pharah",
    "score": 0,
    "you": "Pharah"
  },
  Object {
    "enemy": "Tracer",
    "score": -1,
    "you": "Pharah"
  },
  Object {
    "enemy": "Soldier: 76",
    "score": -1,
    "you": "Pharah"
  },
  Object {
    "enemy": "Roadhog",
    "score": -1,
    "you": "Pharah"
  },
  Object {
    "enemy": "McCree",
    "score": -2,
    "you": "Pharah"
  }
]
`;

exports[`undersight getTopScores should calculate score for four enemy 1`] = `
Array [
  Object {
    "name": "Soldier: 76",
    "score": 8
  },
  Object {
    "name": "Reaper",
    "score": 4
  },
  Object {
    "name": "Pharah",
    "score": 0
  }
]
`;

exports[`undersight getTopScores should calculate score for one enemy 1`] = `
Array [
  Object {
    "name": "Soldier: 76",
    "score": 2
  },
  Object {
    "name": "Reaper",
    "score": 1
  },
  Object {
    "name": "Pharah",
    "score": 0
  }
]
`;

exports[`undersight getTopScores should calculate scores 1`] = `
Array [
  Object {
    "name": "Soldier: 76",
    "score": 6
  },
  Object {
    "name": "Reaper",
    "score": 3
  },
  Object {
    "name": "Pharah",
    "score": 0
  }
]
`;

exports[`undersight getTopScores should calculate scores 2`] = `
Array [
  Object {
    "name": "Pharah",
    "score": 3
  },
  Object {
    "name": "Soldier: 76",
    "score": 0
  },
  Object {
    "name": "Reaper",
    "score": 0
  }
]
`;

exports[`undersight getTopScores should calculate scores 3`] = `
Array [
  Object {
    "name": "Reaper",
    "score": 3
  },
  Object {
    "name": "Soldier: 76",
    "score": 0
  },
  Object {
    "name": "Pharah",
    "score": -3
  }
]
`;

exports[`undersight getTopScores should calculate scores 4`] = `
Array [
  Object {
    "name": "Soldier: 76",
    "score": 2
  },
  Object {
    "name": "Reaper",
    "score": 2
  },
  Object {
    "name": "Pharah",
    "score": 0
  }
]
`;

exports[`undersight getTopScores should calculate scores 5`] = `
Array [
  Object {
    "name": "Reaper",
    "score": 2
  },
  Object {
    "name": "Soldier: 76",
    "score": 2
  },
  Object {
    "name": "Pharah",
    "score": 0
  }
]
`;
